src/auth/linking.ts

```diff
@@ -0,0 +1,299 @@
+  if (!ls.temporaryRsaPair || !ls.temporaryRsaPair.privateKey) return null
+
+  if (ls.sessionKey) {
+    throw new Error("Already got a session key")
+  }
+
+  const json = JSON.parse(data)
+  const iv = utils.base64ToArrBuf(json.iv)
+
+  const encryptedSessionKey = utils.base64ToArrBuf(json.sessionKey)
+  const rawSessionKey = await rsa.decrypt(encryptedSessionKey, ls.temporaryRsaPair.privateKey)
+  const sessionKey = await aes.importKey(utils.arrBufToBase64(rawSessionKey), { alg: SymmAlg.AES_GCM, length: 256 })
+
+  ls.sessionKey = sessionKey
+  ls.temporaryRsaPair = null
+
+  // Extract UCAN
+  const encodedUcan = await aes.decrypt(
+    json.msg,
+    ls.sessionKey,
+    {
+      alg: SymmAlg.AES_GCM,
+      iv: iv
+    }
+  )
+
+  const decodedUcan = ucan.decode(encodedUcan)
+
+  if (await ucan.isValid(decodedUcan) === false) {
+    throw new Error("Invalid closed UCAN")
+  }
+
+  if (decodedUcan.payload.ptc) {
+    throw new Error("Invalid closed UCAN: must not have any potency")
+  }
+
+  const sessionKeyFromFact = decodedUcan.payload.fct[0] && decodedUcan.payload.fct[0].sessionKey
+  if (!sessionKeyFromFact) {
+    throw new Error("Session key is missing from closed UCAN")
+  }
+
+  const sessionKeyWeAlreadyGot = utils.arrBufToBase64(rawSessionKey)
+  if (sessionKeyFromFact !== sessionKeyWeAlreadyGot) {
+    throw new Error("Closed UCAN session key does not match the one we already have")
+  }
+
+  return new Uint8Array(utils.randomBuf(6)).map(n => {
+    return n % 10
+  })
+}
+
+/**
+ * CONSUMER: NEGOTIATION (response)
+ * 
+ * Encrypt and publish the CONSUMER did and generated pin for verification by the PRODUCER.
+ * 
+ * @param pin 
+ * @returns 
+ */
+export const sendUserChallenge = async (pin: Uint8Array): Promise<void> => {
+  if (!ls.sessionKey) return
+  
+  const iv = utils.randomBuf(16)
+  const msg = await aes.encrypt(JSON.stringify({
+    did: await did.ucan(),
+    pin
+  }), ls.sessionKey, { iv, alg: SymmAlg.AES_GCM })
+
+  await publishOnChannel(
+    JSON.stringify({
+      iv: utils.arrBufToBase64(iv),
+      msg
+    })
+  )
+  nextStep()
+}
+
+/**
+ * PRODUCER: NEGOTIATION
+ * 
+ * PRODUCER receives the DID & challenge PIN from the CONSUMER.
+ * 
+ * @param data 
+ * @returns 
+ */
+export const handleUserChallenge = async (data: any): Promise<string | null> => {
+  if (!ls.sessionKey) return null
+
+  const { iv, msg } = JSON.parse(data)
+
+  console.debug("msg: " + msg)
+  if (!iv) {
+    throw new Error("I tried to decrypt some data (with AES) but the `iv` was missing from the message")
+  }
+
+  console.debug("decrypting msg")
+  return await aes.decrypt(msg, ls.sessionKey, {
+    alg: SymmAlg.AES_GCM,
+    iv
+  })
+  nextStep()
+}
+
+/**
+ * PRODUCER: DELEGATION
+ * 
+ * This step is user initiated, create the UCAN with delegate rights for the CONSUMER (and a WNFS read key).
+ * 
+ * @param data 
+ * @returns 
+ */
+export const sendReadKeyAndUcan = async (data: any): Promise<null> => {
+  return null
+}
+
```

Comment:
This is what should get called when the user accepts the PIN on the producer side. The auth-lobby code lives here: https://github.com/fission-suite/auth-lobby/blob/develop/src/Javascript/Main.js#L783

Because this is "user facing" (i.e. the implementor needs to trigger this in response to user action)... I'm not sure the best way to surface this. Maybe a friendlier / clearer name is all it needs?

VERY open to thoughts here.